### DISCUSSION TITLE
Write-Output cannot get as an inputobject a ScriptBlock

### DISCUSSION THREAD

**Discussion by RobertoRojas** [PowerShell/PowerShell#16560] [Q&A] [answered]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Hello everyone, I'm working with PWSH, the idea in general is have a ps1 file with the definition of several scriptblocks and select one (to have control of the version to execute). To get the block I'm using the function Write-Output to return it. But it throw a weir error. 

This problem only happen with the ScriptBlock, I checked the expected object of Write-Output:

Write-Output [-InputObject] <psobject> [-NoEnumerate] [<CommonParameters>]

@{}.GetType() -> System.Object
{}.GetType() -> System.Object

Is possible write a hashtable, and the both are System.Object.

Obviously, you can have different workarounds:

```powershell
Invoke-Command -ScriptBlock $([Scriptblock]::Create((Write-Output -InputObject {Write-Host 'Hello block!'}.ToString())));
```

```powershell
Invoke-Command -ScriptBlock $(Write-Output -InputObject $([psobject]@{"Command"={Write-Host 'Hello block!'}})).Command
```

### Expected behavior

```console
PS> Write-Output -InputObject {write-host 'Hello world!'}
write-host 'Hello world!'
```

### Actual behavior

```console
PS> Write-Output -InputObject {write-host 'Hello world!'}
Write-Output: Cannot evaluate parameter 'InputObject' because its argument is specified as a script block and there is no input. A script block cannot be evaluated without input.
```

### Error details

```console
Write-Output: Cannot evaluate parameter 'InputObject' because its argument is specified as a script block and there is no input. A script block cannot be evaluated without input.
```

### Environment data

```powershell
PS> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.2.0
PSEdition                      Core
GitCommitId                    7.2.0
OS                             Linux 5.14.18-100.fc33.x86_64 #1 SMP Fri Nov 12 17:38:44 UTC 2021
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by iSazonov** [comment]
Please reword your OP. Can you use an translator (google, deepl) to get more clear English? Please add more clear repro steps.

**Comment by RobertoRojas** [comment, marked as answer]
Basically, Write-Output with a ScriptBlock as 'InputObject' return the follow error message: 

```Powershell
PS> Write-Output -InputObject {}
Write-Output: Cannot evaluate parameter 'InputObject' because its argument is specified as a script block and there is no input. A script block cannot be evaluated without input.
```

This only happen with the ScriptBlock, you can use 'return' to finish the execution of the script and it will return the ScriptBlock object.

**Reply by SeeminglyScience** [reply]
This is due to how [delay-bind script block parameters](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_script_blocks?view=powershell-7.2#using-delay-bind-script-blocks-with-parameters) work. You can do `0..10 | Write-Output -InputObject { $_ * 2 }`. So typically it's assumed that if you are sending a script block to a command like this then you probably meant to use delay-bind and the pipeline position is a mistake.

This is the case for any parameter that is typed specifically as `PSObject` and is marked as taking input from the pipeline.

**Reply by RobertoRojas** [reply]
Yes, that explain the error message. At the end I used this workarround to provide this behaivor:

Send ScriptBlock
```Powershell
Write-Output -InputObject @{
    "1.0.0" = {
        # Code
    };
}[$Version].ToString();
```

Receive ScriptBlock
```Powershell
[ScriptBlock]::Create($(. "Script path" -Version "1.0.0"));
```

**Reply by SeeminglyScience** [reply]
You can just not use `Write-Output`. e.g. `{ code }` and let it implicitly emit. Or `return { code }`. Or even do all that but just don't stringify it and recreate.

**Reply by RobertoRojas** [reply]
The way that I found to use the Write-Output and ScriptBlocks is the follow:

```Powershell
{write-host 'x'} | Write-Output
```

Basically using the pipe it have the expected behaivor. My problem with the 'return' is the follow:

**Case 1:**
```Powershell
Invoke-Command -ScriptBlock {"A";return;}
```
```Output
A
```

**Case 2:**
```Powershell
Invoke-Command -ScriptBlock {return "A";}
```
```Output
A
```

**Case 3:**
```Powershell
Invoke-Command -ScriptBlock {"A";}
```
```Output
A
```

In all the cases is the same, the keyword works fine in the case that I need to break the execution of the script in the middle:

**Case 4:**
```Powershell
Invoke-Command -ScriptBlock {"A";if($true){return;}"B"}

```
```Output
A
```

**Reply by SeeminglyScience** [reply]
> In all the cases is the same

Yeah, though the same is true for `{ code }` vs `{ code } | Write-Output`.
